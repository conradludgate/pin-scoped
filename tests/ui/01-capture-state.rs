#![feature(async_closure)]

use std::{pin::pin, sync::Mutex};
use tokio::task::yield_now;

use pin_scoped::Scoped;

#[tokio::main]
async fn main()  {
    let mut scoped = pin!(Scoped::new(Mutex::new(0)));

    let non_scoped = Mutex::new(0);

    for _ in 0..64 {
        scoped.as_mut().spawn(async |state: &Mutex<u64>| {
            yield_now().await;
            *state.lock().unwrap() += 1;
            *non_scoped.lock().unwrap() += 1;
            yield_now().await;
        });
    }

    scoped.await;
}
