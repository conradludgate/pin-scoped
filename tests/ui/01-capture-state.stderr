error[E0597]: `non_scoped` does not live long enough
  --> tests/ui/01-capture-state.rs:15:31
   |
15 |            scoped.as_mut().spawn(async |state: &Mutex<u64>| {
   |  __________-_____________________^
   | | _________|
   | ||
16 | ||             yield_now().await;
17 | ||             *state.lock().unwrap() += 1;
18 | ||             *non_scoped.lock().unwrap() += 1;
19 | ||             yield_now().await;
20 | ||         });
   | ||_________^- argument requires that `non_scoped` is borrowed for `'static`
   | |__________|
   |            borrowed value does not live long enough
...
24 |    }
   |    - `non_scoped` dropped here while still borrowed
